### ISSUE TITLE
elm-ls fails to start with 'permission denied'

### ISSUE THREAD

**Issue by teu5us** [emacs-lsp/lsp-mode#1111]
When I open a .elm file, lsp starts, but it cannot connect to elm-language-server with the following in the echo area:

`LSP :: Request initialized failed with message: EACCES: permission denied, scandir '/path/to/some/dir'`

The directory in question is some random directory in my home directory. Why is it even trying to scan everything?

**Comment by teu5us** [comment]
This example from elm website or any other .elm file. Whenever I open an elm file in emacs, lsp cannot initialize. I expect lsp to start, it reports starting the elm-ls in the modeline after the mentioned error, so I get no completion or anything.

Arch Linux, elm-language-server installed from npm.

However, elm-ls works fine with coc.nvim.

```
module Main exposing (Model, Msg(..), init, main, update, view, viewInput, viewValidation)

import Browser
import Html exposing (..)
import Html.Attributes exposing (..)
import Html.Events exposing (onInput)



-- MAIN


main =
    Browser.sandbox { init = init, update = update, view = view }



-- MODEL


type alias Model =
    { name : String
    , password : String
    , passwordAgain : String
    , age : String
    }


init : Model
init =
    Model "" "" "" ""



-- UPDATE


type Msg
    = Name String
    | Password String
    | PasswordAgain String
    | Age String


update : Msg -> Model -> Model
update msg model =
    case msg of
        Name name ->
            { model | name = name }

        Password password ->
            { model | password = password }

        PasswordAgain password ->
            { model | passwordAgain = password }

        Age age ->
            { model | age = age }



-- VIEW


view : Model -> Html Msg
view model =
    div []
        [ viewInput "text" "Name" model.name Name
        , viewInput "number" "Age" model.age Age
        , viewInput "password" "Password" model.password Password
        , viewInput "password" "Re-enter Password" model.passwordAgain PasswordAgain
        , viewAgeValidation model
        , viewValidation model
        ]


viewInput : String -> String -> String -> (String -> msg) -> Html msg
viewInput t p v toMsg =
    input [ type_ t, placeholder p, value v, onInput toMsg ] []


viewAgeValidation : Model -> Html msg
viewAgeValidation model =
    if then
        div [ style "color" "green" ] [ text "Age is correct" ]

    else
        div [ style "color" "red" ] [ text "Enter correct age" ]


viewValidation : Model -> Html msg
viewValidation model =
    if
        String.length model.password
            >= 8
            && String.any Char.isDigit model.password
            && String.any Char.isUpper model.password
    then
        if model.password == model.passwordAgain then
            div [ style "color" "green" ] [ text "OK" ]

        else
            div [ style "color" "red" ] [ text "Passwords do not match!" ]

    else
        div [ style "color" "red" ] [ text "Password should be longer than 8 characters" ]
```
I get nothing except for the following as errors:

```
LSP :: Connected to [elm-ls:287473 status:starting].
LSP :: Request initialize failed with message: EACCES: permission denied, scandir '/home/paul/.acrobat/dosdevices/z:/lost+found'
```

**Comment by yyoncho** [comment]
Hi @Daniel-V1 - I noticed that you are the latest committer in elm related code - have you ever seen anything like described in the PR?

**Comment by teu5us** [comment]
I actually found only one comment about the issue on the internet. So, as I said, elm-ls works fine with coc.nvim and I also tried it with vscode where it also works as expected. I get this error in emacs with both lsp-mode and eglot.

**Comment by yyoncho** [comment]
I am struggling to make elmls work on my side at all (it starts but it errors out that on each call it, fails in vscode too). I believe that the issue is that lsp-mode does not send some information in the initialization options, so capturing the logs from vscode and emacs and comparing both will help us resolve the issue. 

1. In vscode, can you do the following: `C-S-p User Settings RET` and type `Elm Trace`, change it to `verbose` and then View -> Output and select Elm from the drop-down. But before all, make sure that you are using `ElmLS` extension.
2. In emacs: set lsp-print-io to t, restart the server and you could use lsp-workspace-show-log to view the server log.

**Comment by teu5us** [comment]
@yyoncho I'll provide the logs asap, can also try to get logs from coc.nvim for comparison

**Comment by teu5us** [comment]
Here's the lsp log:
[emacslog.txt](https://github.com/emacs-lsp/lsp-mode/files/3746497/emacslog.txt)

This one is from vscode:
[vscodelog.txt](https://github.com/emacs-lsp/lsp-mode/files/3746498/vscodelog.txt)

This is from coc.nvim:
[cocnvimlog.txt](https://github.com/emacs-lsp/lsp-mode/files/3746503/cocnvimlog.txt)

**Comment by teu5us** [comment]
Changing these two lines (on line 5377) in lsp-mode.el got elm-ls running.

*From*:
```
                                :rootPath (lsp-file-local-name (expand-file-name root))
                                :rootUri (lsp--path-to-uri root)
```

*To*:
```
                                :rootPath (lsp-file-local-name (expand-file-name (file-name-directory (buffer-file-name))))
                                :rootUri (lsp--path-to-uri (file-name-directory (buffer-file-name)))
```

However now there is a new issue with the following in the echo area:
```
LSP :: unhandled method textDocument/hover
LSP :: unhandled method textDocument/codeAction
```

**Comment by yyoncho** [comment]
You have selected incorrect project root - 

```
  "rootPath": "/home/paul/",
  "rootUri": "file:///home/paul/",
```

If you have set lsp-auto-guess to t - remove that setting. I recommend not setting this to t until you understand how lsp-mode internals work.

Otherwise, do lsp-workspace-folder-remove and next time when you open the file and call lsp you will be asked to select root. Follow the instructions and select the correct project root.

> LSP :: unhandled method textDocument/hover
> LSP :: unhandled method textDocument/codeAction

I am getting these too, also in vscode but it is eating the exceptions - I think these should be reported in the language server repository. The problem is that the server manifests that it supports hoverProvider but when you call the corresponding it reports that the method is not supported.

**Comment by teu5us** [comment]
lsp-auto-guess-root is set to nil in my case. Whatever file I open gets my homedir as root. The rootPath for elm-ls should be that containing elm.json.

I moved my file Main.elm from src dir to project root dir and with my changes to lsp-mode.el elm-ls started. I get no completion, though.

**Comment by yyoncho** [comment]
> Whatever file I open gets my homedir as root. 

That is because you have added the homedir as a root folder. Call `M-x` `lsp-workspace-folders-remove` first.

> I get no completion, though.

can you try executing this:

```
(setf (lsp--client-multi-root (gethash 'elm-ls lsp-clients)) t)
```

note that you should do that before starting the elm language server.

If it doesnt work, I will need the log from vscode from opening the file to the point at which you receive completion and the same thing for emacs.

**Comment by teu5us** [comment]
I see. I removed homedir as you said and added the required dir as root. Starts as expected. Completion works, but only through M-Tab

**Comment by yyoncho** [comment]
Do you have company-lsp installed?

**Comment by yyoncho** [comment]
Also, does it work without `(setf (lsp--client-multi-root (gethash 'elm-ls lsp-clients)) t)` ? If it requires that option, I will update the default configuration to have multi-folder enabled by default.

**Comment by teu5us** [comment]
I do have that one installed, works with other backends. Completion works without that option.

**Comment by yyoncho** [comment]
Does it work when you explicitly do `M-x` `company-lsp`? Also, what is the value of `company-backends` when you are in elm buffer?

**Comment by teu5us** [comment]
Yes, works when told explicitly.
`company-backends` is `(company-elm (company-semantic company-dabbrev-code company-gtags company-etags company-keywords) company-files company-dabbrev)`

**Comment by yyoncho** [comment]
@teu5us apparently you have something in your config that removes company-lsp from company-backends. I believe that company-elm is doing 

Something like that will fix it, but it should be executed after company-elm action.

```
(add-hook 'elm-mode-hook (lambda () (setq-local company-backends '(company-lsp))))
```
 Ideally, you should find a way to disable that package

**Comment by teu5us** [comment]
I guess that's some issue on spacemacs side.

This one did not help, unfortunately.
```
(add-hook 'elm-mode-hook (lambda () (setq-local company-backends '(company-lsp))))
```

**Comment by yyoncho** [comment]
Delete this from the elm layer and it will work. 

```
(defun elm/post-init-company ()
  (spacemacs|add-company-backends :backends company-elm :modes elm-mode)
  (add-hook 'elm-mode-hook 'elm-oracle-setup-completion))
```

**Comment by teu5us** [comment]
company also needs to be removed from elm-packages, otherwise there's a  warning. It would be nice to have lsp-elm as an option like they have in their python layer.

I changed company-elm to company-lsp and the hook to elm-oracle to lsp. Both ways it works, but quite slowly. Anyway, thanks for your help
